Add CLIs for simple additive bias correction

benowen-bom opened the pull request:
**Apologies for the length of this PR:** there was a rush on the BoM to get this in quickly, so I haven't had the time (in part due to Ops commitments) to break this down into smaller PRs.

This PR adds functionality to enable the application of a simple additive bias correction process applied on a per member basis using forecast bias values from a representative single value forecast.

The functionality is separated across two separate Plugins and associated CLIs:

1. Calculate forecast bias
2. Apply bias correction

Here we include additive error only, however the code has been structured such that expanding the usage to other methods (such as multplicative bias) could be included realtively simply.

The usage here is intended to enable evaluation and application of bias as either:
- multiple historical forecasts bundled into a single forecast bias file (storing mean value),
- forecast bias is evaluated on individual forecasts separately, then passed in as a list in the apply-bias-correction step
This choice is to enable flexibility for other methods (which may not allow for evaluation on per forecast basis), while aligning with how other processes (such as rel cal) operator to enable similar behaviour on the suite end.

The additions are the following:
- [x] CLI to calculate forecast bias (`improver/cli/calculate_forecast_bias.py`)
- [x] ClI to apply bias correction (`improver/cli/apply_forecast_bias.py`)
- [x] Plugins to perform evaluation and application of additive forecast error (bias)
- [x] Associated unit tests
- [x] Acceptance tests

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

btrotta-bom reviewed (commented):
The logic looks ok, I just had a couple of minor comments about docstrings etc.

benowen-bom commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,284 @@
+    forecasts: Cube, truths: Cube, collapse_dim: str
+) -> ndarray:
+    """
+    Evaluate the mean additive error between the forecast and truth dataset.
```
Good suggestion. I've added this to the associated doc-string.

benowen-bom commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,284 @@
+
+def apply_additive_correction(forecast: Cube, bias: Cube) -> ndarray:
+    """
+    Apply additive correction to forecast using the specified bias values.
```
Done.

benowen-bom commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,284 @@
+        Evaluate forecast bias over the set of historic forecasts and associated
+        truth values.
+
+        Where mulitple forecasts values are provided, forecasts must have consistent
```
Fixed.

benowen-bom commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,284 @@
+        historic_forecasts, truths = filter_non_matching_cubes(
+            historic_forecasts, truths
+        )
+        # Ensure that input forecasts are for consitent period/valid-hour
```
Fixed.

benowen-bom commented on improver/cli/apply_bias_correction.py:
```diff
@@ -0,0 +1,78 @@
+    from improver.calibration.simple_bias_correction import ApplyBiasCorrection
+
+    # Check whether bias data supplied, if not then return unadjusted input cube.
+    if not bias_cubes:
+        return forecast_cube
```
This is something that @nivnac requested. The reason for this behaviour is to allow for spin-up of the bias correction terms, so until these exist the default behaviour is to do nothing.

I've added a comment to explain this.

benowen-bom commented on improver/cli/calculate_forecast_bias.py:
```diff
@@ -0,0 +1,65 @@
+    Args:
+        cubes (list of iris.cube.Cube):
+            A list of cubes containing the historical forecasts and corresponding
+            truths used for calibration. The cubes much include the same diagnostic
```
Fixed.

benowen-bom commented on improver/cli/calculate_forecast_bias.py:
```diff
@@ -0,0 +1,65 @@
+            truths used for calibration. The cubes much include the same diagnostic
+            name in their names. The cubes will be distinguished using the user
+            specified truth attribute.
+        truth_attribite (str):
```
Fixed.

benowen-bom commented on improver_tests/acceptance/test_apply_bias_correction.py:
```diff
@@ -0,0 +1,109 @@
+    """
+    Test case where lower bound is supplied.
+
+    Note: we are using an artificially higher lower bound that would use
```
Fixed.

benowen-bom commented:
> The logic looks ok, I just had a couple of minor comments about docstrings etc.

Thanks for your review @btrotta-bom. I've addressed all your comments.

However, @nivnac has raised a question around the behaviour when there are masked values in the `bias_values`. At present the default is to carry the masked areas across, however I think the desired (default behaviour) would be to set these to zero and preserve the data in the masked areas as is. I'm going to incorporate this addition next, so may be worth another look once the associated updates (and unit tests) are in.

nivnac commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,300 @@
+        """
+        attributes = self._define_metadata(forecasts)
+        forecast_bias_cube = create_new_diagnostic_cube(
+            name=f"{forecasts.name()}_forecast_error",
```
I notice that Rainforests and ensemble_copula_coupling uses the name `forecast_error_of_*` rather than `*_forecast_error`.

benowen-bom replied:
Thanks @nivnac. I had a chat with Tim around what cf-conventions say we should do on this, but failed to look back over what we had used for rainforests work. I think being consistent is important, so I'll update the variable naming across the PR.

benowen-bom replied:
Also, do you think we need any other metadata or attributes?

nivnac replied:
None that I can think of.

benowen-bom replied:
Great, thanks James! I'm just updating the acceptance test data and then I think this should be ready for MO review.

benowen-bom commented:
Thanks @btrotta-bom and @nivnac. I'll add the tag now to get MO to review.

gavinevans commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,300 @@
+        Returns:
+            Bias corrected forecast cube.
+        """
+        bias = self._get_mean_bias(bias)
+
+        corrected_forecast = forecast.copy()
+        corrected_forecast.data = self.correction_method(
+            forecast, bias, fill_masked_bias_values
+        )
```
Just wondering, if there's any concern about applying the bias to inappropriate forecasts? At the moment, there's no checking of e.g. forecast_reference_time or forecast_period, so any bias could be applied to any forecast. Is that flexibility intentional or an oversight?

benowen-bom replied:
I've add a function to check the inputs prior to doing the bias-correction. This function checks that all bias datasets are consistent (where multiple are passed in) and that the forecast and bias data are consistent. Where either the `forecast_reference_time` valid-hour or `forecast_period` are different across datasets, a `ValueError` is raised.

btrotta-bom replied:
Is there a case where we might want to combine different forecast reference times? E.g. if we have only a short period of data and want to increase the sample size, and we think the bias is mostly due to location features and would be consistent over different forecast reference times. If we may want to use it this way, we could raise a warning instead of error.

btrotta-bom replied:
The flag sounds like a reasonable solution to me.

benowen-bom commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,300 @@
+        # Collapse the time values down to a single value as mean value
+        # will be stored where multiple forecast_reference_times are passed in.
+        if "time" in get_dim_coord_names(forecast_bias_cube):
+            frt_coord = create_unified_frt_coord(
+                forecast_bias_cube.coord("forecast_reference_time")
+            )
```
Yeah, this thought had occurred to me but in the interests of time left the behaviour as is. One option here could be to include an attribute to specify how many forecast_reference_times have been included in the bias calculation; this coupled with the bounds will highlight that there were missing frts.

gavinevans replied:
Yes, to be fair, I don't think that we account for this possibility anywhere ourselves, so there's isn't any existing agreed metadata to account for this. I'm happy to leave this as it is for now but we could consider it in the future, as we could update this behaviour in other calibration approaches e.g. EMOS and reliability calibration, as well.

benowen-bom commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,300 @@
+            for bias_cube in bias_values:
+                if bias_cube.coord("forecast_reference_time").bounds is not None:
+                    raise ValueError(
+                        "Collapsing multiple bias values to a mean value is unsupported for "
+                        "bias values defined over multiple reference forecast values."
```
I've updated the comment to mention the frt and associated bounds for the bias cube which raises the error.

benowen-bom commented on improver_tests/calibration/simple_bias_correction/test_ApplyBiasCorrection.py:
```diff
@@ -0,0 +1,289 @@
+
+    # Collapse down frt coord if using mean bias value defined in single cube.
+    if single_frt_with_bounds and num_frts > 1:
+        bias_cubes = bias_cubes.merge_cube()
+        frt_coord = create_unified_frt_coord(
+            bias_cubes.coord("forecast_reference_time")
+        )
+        bias_cubes = collapsed(
+            bias_cubes, "forecast_reference_time", iris.analysis.MEAN
+        )
+        bias_cubes.data = bias_cubes.data.astype(bias_cubes.dtype)
+        bias_cubes.replace_coord(frt_coord)
+        bias_cubes = CubeList([bias_cubes])
```
Yeah, this is worthwhile change. I think as it is, line 152 isn't actually doing anything worthwhile (the intention is to ensure the output remains `np.float32`). The reasoning for using `bias_cubes` was that I didn't want to have multiple return statements, but your solution is clearer and respects that so I'll update the code accordingly.

benowen-bom commented on improver_tests/calibration/simple_bias_correction/test_ApplyBiasCorrection.py:
```diff
@@ -0,0 +1,289 @@
+    )
+    with pytest.raises(ValueError):
+        ApplyBiasCorrection()._get_mean_bias(input_cubelist)
+    # Set up cube inputs with mixed definition: one over single frt, the other over
+    # multiple frt values.
+    input_cubelist = CubeList()
+    input_cubelist.extend(generate_bias_cubelist(2, single_frt_with_bounds=False))
+    input_cubelist.extend(
+        generate_bias_cubelist(
+            2,
+            single_frt_with_bounds=True,
+            last_valid_time=VALID_TIME - timedelta(days=2),
+        )
+    )
+    with pytest.raises(ValueError):
+        ApplyBiasCorrection()._get_mean_bias(input_cubelist)
```
Good suggestion. This will key is to test the case where two cubes have bounds associated and where one does and the other doesn't. Both these case can be accommodated with the same code using the `@pytest.mark.parametrize` as suggested.

benowen-bom commented on improver_tests/calibration/simple_bias_correction/test_CalculateForecastBias.py:
```diff
@@ -0,0 +1,223 @@
+def test_process(num_fcst_frt, num_truth_frt):
+    """Test process function over a variations in number of historical forecasts and
+    truth values passed in."""
+    reference_forecast_cubes = generate_dataset(num_fcst_frt)
+    truth_cubes = generate_dataset(num_truth_frt, truth_dataset=True)
+    result = CalculateForecastBias().process(reference_forecast_cubes, truth_cubes)
```
Yeah, you're right here. The return type for `generate_dataset` is a `Cube` rather than a `CubeList`. I've updated this instance a few others above.

benowen-bom commented on improver/calibration/simple_bias_correction.py:
```diff
@@ -0,0 +1,300 @@
+        self,
+        forecast: Cube,
+        bias: CubeList,
+        lower_bound: Optional[float],
```
I've added in a `upper_bound` argument now in addition to `lower_bound`. I've also changed the clipping from using `np.ma.where` to `np.ma.clip`.

btrotta-bom reviewed (dismissed):
This looks fine, I had one comment/question for consideration, but it's optional.

gavinevans commented on improver_tests/calibration/simple_bias_correction/test_ApplyBiasCorrection.py:
```diff
@@ -220,24 +219,71 @@ def test_get_mean_bias_fails_on_inconsistent_bounds():
+
+@pytest.mark.parametrize("num_bias_inputs", (1, 5))
+def test_inconsistent_bias_forecast_inputs(forecast_cube, num_bias_inputs):
+    """Test the check to see whether inputs have consistent forecast_refernce_time
```
```suggestion
    """Test the check to see whether inputs have consistent forecast_reference_time
```

btrotta-bom replied:
Fixed

gavinevans commented on improver_tests/acceptance/test_apply_bias_correction.py:
```diff
@@ -0,0 +1,162 @@
+    acc.compare(output_path, kgo_path)
+
+
+def test_lower_bound(tmp_path):
```
Perhaps this should be `test_bounds`, as both the lower and upper bounds are tested?

btrotta-bom replied:
Fixed

gavinevans commented on improver_tests/acceptance/test_apply_bias_correction.py:
```diff
@@ -0,0 +1,162 @@
+    than would use in practice to ensure process works.
+    """
+    kgo_dir = acc.kgo_root() / "apply-bias-correction"
+    kgo_path = kgo_dir / "with_lower_bound" / "kgo.nc"
```
Perhaps this should now be:
```suggestion
    kgo_path = kgo_dir / "with_bounds" / "kgo.nc"
```

btrotta-bom replied:
Updated

gavinevans reviewed (changes requested):
Thanks for the updates, @benowen-bom 👍 

I've added some very minor comments.

btrotta-bom commented:
Thanks @gavinevans. Since Ben is away for most of January, he's asked me to progress this PR on his behalf. I've made the changes you suggested.

gavinevans reviewed (approved):
Thanks for the updates, @btrotta-bom 👍 

I'm happy with this PR as it stands, although please note that I haven't run the acceptance tests for the new CLIs, as I don't think that I've got access to the new data. With regards to the leniency suggested in https://github.com/metoppv/improver/pull/1843#discussion_r1053889630, is the intention to add this functionality as part of further work?

btrotta-bom commented:
@gavinevans Yes, we don't need this functionality at the moment so we could do it as a later PR if needed.

benowen-bom commented:
Thanks @gavinevans and @btrotta-bom for following this up while I was away.
